### PR TITLE
feat: propose a pattern for shared tool usage

### PR DESCRIPTION
**Pull request by anandaroop**
https://www.notion.so/artsy/Reusable-Tool-Library-0c3660b14a5446dc87ba7629a2cb5786

As we continue to incorporate function calling / tool usage into our experiments, I propose that we develop some stronger patterns for tool usage so that we can:

- build up a reusable tool library with conventions for style, naming, etc
- improve ergonomics and readability
- enjoy better dev-time and run-time type safety for tool use

To accomplish that, this PR brings Vercel's AI SDK into Quantum.

You'll recall that I took a swing at using this library early on, and [ran away in tears](https://artsy.slack.com/archives/C06SSV1K10D/p1713886139485229?thread_ts=1713794399.695939&cid=C06SSV1K10D). That was because we were trying to use its bleeding edge frontend capabilities (i.e. all the "generative UI" / streaming compontents / RSC stuff). I still stand by our decision to stay away from all _that_ for the foreseeable future.

Meanwhile, however, this library has been maturing esp with regard to its vendor- and framework-agnostic bits.

In particular, I noticed that [Vercel absorbed ModelFusion](https://vercel.com/blog/vercel-ai-sdk-3-1-modelfusion-joins-the-team) very recently, which means that they've been incorporating ModelFusion's very nice abstractions into their library.

So here I demonstrate three things:

1. Using the basic features of what Vercel now calls [AI SDK Core](https://sdk.vercel.ai/docs/ai-sdk-core), which I read as the subset of ModelFusion that has been integrated into the SDK.

2. A tool library pattern that makes use of the [tool calling](https://sdk.vercel.ai/docs/ai-sdk-core/tools-and-tool-calling) portion of AI SDK Core, which is the main reason for this PR

3. The ability to use these patterns without lock-in, i.e. easily be able to pass tools defined in this way to the lower level vendor-specific client libraries.

### CODE REVIEW

File: src/09-tool-patterns/03-vanilla-tool-call.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/**
+ * Demonstrates that tools defined via Vercel AI SDK and
+ * can easily be called via the vendor's own API client and
+ * need not be locked into the Vercel abstractions.
```
Comment (anandaroop) [inline src/09-tool-patterns/03-vanilla-tool-call.ts:4]: Once concern when reaching for these abstractions is how heavy and sticky they are, in terms of lock-in, escape hatches etc. So this example was just to prove to myself that the tool abstraction works just fine if we need to break out and supply it directly to one of the vendor's API clients.

File: src/09-tool-patterns/03-vanilla-tool-call.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  // define tools, by accessing the AI SDK tool's attributes
+  // and shaping them for the vendor's API client
+
+  const tools: OpenAI.Chat.ChatCompletionTool[] = [
+    {
+      type: "function",
+      function: {
+        name: "fetchShowsNearLocation",
+        description: fetchShowsNearLocation.description,
+        parameters: zodToJsonSchema(fetchShowsNearLocation.parameters),
```
Comment (anandaroop) [inline src/09-tool-patterns/03-vanilla-tool-call.ts:41]: The `tool` [object returned](https://github.com/artsy/quantum/pull/14/files#diff-3ac74bab8cef087a9f693269aee6a4ebdba8f544bed3911192e2bc011632109cR8) is easily adapted here to use with the OpenAI client directly…

File: src/09-tool-patterns/03-vanilla-tool-call.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+  // handle the tool call
+
+  const toolCall = response.choices[0].message.tool_calls![0]
+
+  console.log("\nTool call:", toolCall.function)
+
+  if (toolCall.function.name === "fetchShowsNearLocation") {
+    const args = JSON.parse(toolCall.function.arguments)
+    const result = await fetchShowsNearLocation.execute(args)
```
Comment (anandaroop) [inline src/09-tool-patterns/03-vanilla-tool-call.ts:63]: …and here for the actual function call

File: src/lib/tools/fetchShowsNearLocation.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      .number()
+      .min(-90)
+      .max(90)
+      .default(DEFAULT_NUMBER_OF_SHOWS)
+      .describe(`The latitude of the location to search near`),
+    size: z
+      .number()
+      .max(MAX_NUMBER_OF_SHOWS)
+      .default(DEFAULT_NUMBER_OF_SHOWS)
+      .describe(`The maximum number of shows to return`),
```
Comment (anandaroop) [inline src/lib/tools/fetchShowsNearLocation.ts:30]: Because of the way that the AI SDK incorporates the Zod schema, it serves nicely as inline documentation, dev-time intellisense _and_ runtime safety. So much winning you will get tired of winning.

An example of user input that causes the model to call a function with disallowed params:

<pre>
InvalidToolArgumentsError [AI_InvalidToolArgumentsError]: Invalid arguments for tool fetchShowsNearLocation: Type validation failed: Value: {"latitude":40.8503876,"longitude":-73.93699079999999,"size":25}.
Error message: [
  {
    <b>"code": "too_big",</b>
    <b>"maximum": 20,</b>
    "type": "number",
    "inclusive": true,
    "exact": false,
    <b>"message": "Number must be less than or equal to 20",</b>
    "path": [
      "size"
    ]
  }
]
</pre>

File: src/lib/metaphysics.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+   */
+  headers: Record<string, string>
+}
+
+/**
+ * Helper to make a request to Metaphysics
+ */
+
+export async function metaphysics(args: Args) {
+  const { query, variables } = args
```
Comment (anandaroop) [inline src/lib/metaphysics.ts:23]: We can re-use this now, instead of reinventing it each time. A good time to add better types and error handling.

File: src/lib/tools/fetchUserArtistFollows.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  }),
+
+  execute: async (args) => {
+    const accessToken = process.env.USER_ACCESS_TOKEN
+    const size = args.size ?? DEFAULT_NUMBER_OF_ARTISTS
+
+    let artistFollows: unknown[] = []
+    let lastCursor = null
+
+    while (artistFollows.length < size) {
```
Comment (anandaroop) [inline src/lib/tools/fetchUserArtistFollows.ts:29]: Building in support for cursor-based pagination so that we can fetch multiple pages of results if needed (still governed by whatever `max` we set, e.g. L15 above)

File: src/lib/tools/index.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+export * from "./fetchUser"
+export * from "./fetchUserArtistFollows"
+export * from "./fetchShowsNearLocation"
```
Comment (anandaroop) [inline src/lib/tools/index.ts:3]: Setting up for easier imports.

I believe this pattern of re-exporting everything from a root file is frowned on in Force due to tree-shaking(?) concerns, so maybe we don't do it there? But seems fine for quantum.

File: src/09-tool-patterns/01-demo.ts [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+ */
+
+const addTwoNumbers = tool({
+  description: "Add two numbers",
+  parameters: z.object({
+    a: z.number(),
+    b: z.number(),
+  }),
+  execute: async ({ a, b }) => Promise.resolve(a + b),
+})
```
Comment (damassi) [inline src/09-tool-patterns/01-demo.ts:195]: 👌 Very nice!

File: src/lib/metaphysics.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
+interface Args {
```
Comment (damassi) [inline src/lib/metaphysics.ts:1]: Rather than `lib` (which is somewhat foreign at artsy these days), maybe follow some of our other recent JS patterns where we make these types of things first class in a `system` directory. 

So 
- `system/metaphysics.ts` (or `system/network` -- ideally, later, it's `system/relay`, which is what we see everywhere now) 
- `system/tools`

File: src/lib/tools/index.ts (hunk 7 above)
Comment (damassi) [reply src/lib/tools/index.ts:3]: Please avoid these types of index exports! Even though this is an experimental app, we don't know where it will finally go. Its impossible to express how much time and effort has gone into basically (all) of our apps unwinding this index pattern; The confusion it generates coupled with VSCode auto import is great, and it makes refactoring opaque. Please trust me 🙏

File: src/09-tool-patterns/01-demo.ts [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+  /*
+   * Simplest demo
+   */
+
+  let response = await generateText({
+    model: openai("gpt-4o"),
+    temperature: 0,
+    system,
+    prompt,
+  })
```
[the remark this answers is not in the document]
Comment (mc-jones) [reply src/09-tool-patterns/01-demo.ts:43]: I like the separation of the system instructions and the prompt/messages.

Do the elements in a `messages` `array` have the same shape (i.e. `[{role: 'foo', content: 'bar'}]`)?

[the remark this answers is not in the document]
Comment (mc-jones) [reply src/09-tool-patterns/01-demo.ts:43]: See the code further down, the answer looks to be yes, and I would assume this applies regardless of model provider.

File: src/lib/tools/fetchUser.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+            lng
+          }
+        }
+      }
+    }`
+
+    const variables = {}
+
+    const headers = {
+      "X-ACCESS-TOKEN": accessToken!,
```
Comment (mc-jones) [inline src/lib/tools/fetchUser.ts:40]: I think this is fine for a demo repo/PR, but as we move towards more advances patterns that could end up in production (possibly), just noting we make sure to should be careful this non-null guarantee doesn't get copied over. Would be an easy one to miss.

**Review by mc-jones — APPROVED** [review]
This is great, and I definitely think worth continuing to experiment with. We discussed them when you gave us the walk through, but just to document my primary feedback: 

- We should take some time to test out the browser client before adopting more significantly. 
- I appreciate this [exploration](https://github.com/artsy/quantum/pull/14/files#r1608673180) as lock in with vercel id definitely something we should consider more deeply before full scale adoption.

File: src/09-tool-patterns/01-demo.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (anandaroop) [reply src/09-tool-patterns/01-demo.ts:43]: That's right, a standard message list, typed as `CoreMessage[]`

File: src/09-tool-patterns/02-artsy-tools.ts [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+      content: dedent`
+        Are there any shows near me with artists that I follow?
+      `,
+    },
+  ]
+
+  let currentIteration = 0
+  let isDone = false
+
+  while (!isDone && currentIteration < MAX_ITERATIONS) {
```
[the remark this answers is not in the document]
Comment (anandaroop) [reply src/09-tool-patterns/02-artsy-tools.ts:47]: A `while (true)`-style event loop is pretty standard for this kind of scenario, though? But I'm not denying that I should embrace recursion more 😄.

File: src/lib/metaphysics.ts (hunk 9 above)
Comment (anandaroop) [reply src/lib/metaphysics.ts:1]: Maybe, top-level `tools` makes sense, esp given the other comment below.

File: src/lib/metaphysics.ts (hunk 5 above)
Comment (anandaroop) [reply src/lib/metaphysics.ts:23]: Will definitely need your help connecting the dots if/when we get to that point. Not clear to me how (or even why?) we would use Relay for this kind of server-side imperative code.

File: src/lib/tools/fetchUser.ts (hunk 11 above)
Comment (anandaroop) [reply src/lib/tools/fetchUser.ts:40]: Yeah good call. Probably other improvements to make re: token handling as well, like it should maybe be passed in from the caller, rather than assumed to be in the env.

File: src/lib/tools/index.ts (hunk 7 above)
Comment (anandaroop) [reply src/lib/tools/index.ts:3]: Ha, I was hoping to get your input on this 😄 

Yeah, can definitely back away from this, and I think the imports will also read nicer if they are from a top-level `tools` directory, as suggested above. I def trust you on this, but do you have a handy reference to a discussion about the problems with this approach. I was trying to remember but honestly could not recall the headaches it caused.

**Review by anandaroop — COMMENTED** [review]
Thanks, will address a few comments before merging

File: src/lib/metaphysics.ts (hunk 5 above)
Comment (damassi) [reply src/lib/metaphysics.ts:23]: Generated types (!), caching, familiar API, relay network layer for debugging, and so much more.

File: src/lib/tools/index.ts (hunk 7 above)
Comment (damassi) [reply src/lib/tools/index.ts:3]: I wrote a small summary here: https://artsy.slack.com/archives/C2TQ4PT8R/p1671568357903639

TLDR: 100% of the time, refactors become a real headache. And the gain is zero, compared to just concrete single locations that cannot be interpreted. A file import path is like a memory address. 1:1.

Comment (anandaroop) [reply src/lib/tools/index.ts]: Thanks for that refresher. Unwound in https://github.com/artsy/quantum/pull/14/commits/fab86c8954b80941e3cb3cfb5ebc1e8737ef1e60

File: src/lib/metaphysics.ts (hunk 9 above)
Comment (anandaroop) [reply src/lib/metaphysics.ts]: Also in https://github.com/artsy/quantum/pull/14/commits/fab86c8954b80941e3cb3cfb5ebc1e8737ef1e60

**Comment by anandaroop** [comment]
Tool imports look like this now:

https://github.com/artsy/quantum/blob/fab86c8954b80941e3cb3cfb5ebc1e8737ef1e60/src/09-tool-patterns/02-artsy-tools.ts#L16-L18

I'll pause here for now, but open to discussion in future PRs

File: src/09-tool-patterns/02-artsy-tools.ts [hunk 13]
```diff
@@ -13,11 +13,9 @@ import chalk from "chalk"
 import { CoreMessage, generateText } from "ai"
 import { openai } from "@ai-sdk/openai"
 import dedent from "dedent"
-import {
-  fetchShowsNearLocation,
-  fetchUser,
-  fetchUserArtistFollows,
-} from "../lib/tools"
+import { fetchUser } from "@/tools/fetchUser"
```
Comment (damassi) [inline src/09-tool-patterns/02-artsy-tools.ts:16]: @anandaroop - This can be for a follow-up, but we don't need to use `@` special characters for absolute imports; we can just set it up so that we use the actual folders. Happy to help set this up!

File: tsconfig.json [hunk 14]
```diff
@@ -23,13 +21,19 @@
...
     // "moduleResolution": "node10",                     /* Specify how TypeScript looks up a file from a given module specifier. */
-    // "baseUrl": "./",                                  /* Specify the base directory to resolve non-relative module names. */
-    // "paths": {},                                      /* Specify a set of entries that re-map imports to additional lookup locations. */
+    "baseUrl": "./src", /* Specify the base directory to resolve non-relative module names. */
+    "paths": {
+      "@/tools/*": [
+        "tools/*"
+      ],
+      "@/system/*": [
+        "system/*"
```
Comment (damassi) [inline tsconfig.json:34]: (Also no need to manually assign things)

File: src/09-tool-patterns/02-artsy-tools.ts (hunk 13 above)
Comment (anandaroop) [reply src/09-tool-patterns/02-artsy-tools.ts:16]: Yep I know it's not a requirement. This is a [Next-ism](https://nextjs.org/docs/app/building-your-application/configuring/absolute-imports-and-module-aliases) I have adopted because it clearly signifies which imports are project-relative absolutes and which are from packages. Without the `@` (or equivalent [sigil](https://en.wikipedia.org/wiki/Sigil_(computer_programming)) — though I like @ because of its mnemonic value for <b><i>a</i></b>bsolute imports) it's needlessly ambiguous. I wish we adopted this more widely! 😉

Comment (damassi) [reply src/09-tool-patterns/02-artsy-tools.ts:16]: Its more information in the system which = complexity 🤷

Comment (damassi) [reply src/09-tool-patterns/02-artsy-tools.ts:16]: (There's also no real reason to separate packages from project code, since it doesn't create a problem)
